test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,289 @@
+    std::for_each(keys_begin, keys_begin + n, [&histogram](const auto& key) { ++histogram[key]; });
+    auto first = oneapi::dpl::make_zip_iterator(keys_begin, vals_begin);
+    auto has_duplicates = [&histogram](const auto& pair) {return histogram[std::get<0>(pair)] > 1; };
+    auto new_last = std::remove_if(first, first + n, has_duplicates);
+    return std::distance(first, new_last);
+}
+
+template<typename KeyIt, typename ValIt, typename Size>
+void
+generate_data(KeyIt keys_begin, ValIt vals_begin, Size n, std::uint32_t seed)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    if constexpr (std::is_same_v<ValT, Particle>)
+    {
+        static_assert(std::is_same_v<KeyT, Particle::energy_type>);
+        std::default_random_engine gen{seed};
+        std::uniform_real_distribution<float> mass_dist(0.0, 1000.0);
+        std::uniform_real_distribution<float> velocity_dist(0.0, 1.0);
+        std::uniform_real_distribution<float> coord_dist(0.0, 1.0);
+        std::generate(vals_begin, vals_begin + n, [&gen, &mass_dist, &velocity_dist, &coord_dist]() {
+            return Particle{mass_dist(gen), velocity_dist(gen), {coord_dist(gen), coord_dist(gen), coord_dist(gen)}};
+        });
+        std::transform(vals_begin, vals_begin + n, keys_begin, [](const auto& particle) { return particle.energy(); });
+    }
+    else
+    {
+        TestUtils::generate_arithmetic_data(keys_begin, n, seed);
+        TestUtils::generate_arithmetic_data(vals_begin, n, seed + 1);
+        // avoid having value duplicates at the same locations as key duplicates for a more robust stability check
+        std::shuffle(vals_begin, vals_begin + n, std::default_random_engine(seed));
+    }
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, StableSort)
+{
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSort)
+{
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare)
+{
+    static_assert(std::is_same_v<Compare, DefaultComparator> || std::is_same_v<Compare, GreaterComparator>);
```

Comment:
why only default  and great?
Why not general Compare?

Reply:
Note that a general comparator was not tested originally. This PR extends the testing as it is described, well, in the PR's description.

Do you think that adding a custom comparator is still necessary? It requires some kind of mapping between the comparators for the tested algorithm (`dpl::sort_by_key`) and the checker (`std::sort`).  Probably, it would be better to pass both comparators: one for `dpl::sort_by_key` and another for `std::sort` from `test` functions instead of a single comparator-tag.  

Would you prefer me implementing the idea above, or leaving it as is? I am inclined toward leaving it as is for now.

Reply:
> a general comparator was not tested originally

I means the test was written quite right originally...

As example, I keep in my mind the following simple test workflow:
```
test_sort(sort_algo, compare, auto&&... data)
{
      sort_algo(compare, data...);
      check_result(compare, data...)
}
//.....
//test case 1:
test_sort(sort_stable, std::less{}, values);
//test case 2:
test_sort(sort_unstable, my_comparator, values);
//test case 3:
test_sort(sort_stable_by_key, std::greater{}, values, keys);
//etc
```

Reply:
I've added https://github.com/oneapi-src/oneDPL/issues/1780 for further refactoring.